recaf-launcher/src/main/java/dev/xdark/recaf/Launcher.java

```diff
@@ -0,0 +1,399 @@
+    logger.info("Target jar: {}", jarPath);
+    boolean exists = Files.exists(jarPath);
+
+    logger.info("Target jar exists: {}", exists);
+    if (info == null) {
+      // We were unable to fetch latest release info.
+      // Attempt to continue.
+      if (exists) {
+        attemptLaunch(jarPath, updateFailedFlag("missingInfo"));
+      } else {
+        logger.error("Launcher was unable to fetch release info, cannot continue");
+        logger.error("If you believe that it is a bug, please open an issue at: ");
+        logger.error(ISSUES_URL);
+        System.exit(1);
+      }
+    }
+    boolean autoUpdate = options.valueOf(autoUpdateOption);
+    logger.info("Automatic updates: {}", autoUpdate);
+    boolean versionValid = false;
+    if (exists) {
+      String version = null;
+      try {
+        Manifest manifest = getManifest(jarPath);
+        version = manifest.getMainAttributes().getValue("Specification-Version");
+        logger.info("Detected current version: {}", version);
+      } catch (IOException ex) {
+        logger.error("Unable to get Recaf version: ", ex);
+      }
+      if (version == null) {
+        logger.error("Version detection has failed");
+        logger.error("Ensure that Recaf's jar file is not damaged, or");
+        logger.error("open an issue if you think that it's an error: ");
+        logger.error(ISSUES_URL);
+        try {
+          Path target = jarPath.getParent().resolve(
+              jarPath.getFileName().toString() + "." + System.currentTimeMillis() + ".bak");
+          Files.copy(jarPath, target);
+        } catch (IOException ex) {
+          logger.warn("Could not copy old jar: ", ex);
+        }
+        versionValid = false;
+      } else {
+        try {
+          versionValid = !isOutdated(version, info.getName());
+        } catch (Exception ex) {
+          logger.error("Could not compare versions: ", ex);
+          versionValid = false;
+        }
+      }
+    }
+    Asset asset = findReleaseAsset(info.getAssets());
+    if (asset == null) {
+      logger.error("Unable to detect release asset");
+    }
+    boolean areSizesEqual = false;
+    if (versionValid) {
+      if (asset == null) {
+        logger.error("Launcher was unable to detect release asset from GitHub releases");
+        logger.error("Please open an issue at: ");
+        logger.error(ISSUES_URL);
+        attemptLaunch(jarPath, updateFailedFlag("missingAsset"), 1);
+      }
+      try {
+        areSizesEqual = asset.getSize() == Files.size(jarPath);
+        logger.info("Jar size match: {}", areSizesEqual);
+      } catch (IOException ex) {
+        logger.error("Could not verify jar size, will assume that jar is valid: ", ex);
+      }
+    }
+    if (!areSizesEqual) {
+      if (exists) {
+        boolean writeable = Files.isWritable(jarPath);
+        if (writeable) {
+          // Attempt to write zero-length byte array to verify.
+
+          //noinspection EmptyTryBlock
```

Comment:
Got me curious if `isWriteable` is ever wrong, and it looks like that used to happen in JDK 7 but has since been fixed: https://stackoverflow.com/questions/12688287/java-nio-file-files-iswriteable-doesnt-agree-with-java-io-file-canwrite

Its there a reason to do this sanity check?

Reply:
This check is added because isWriteable does not detect whether file is opened.
For example, by 7-Zip or any other programs.

Reply:
Oh good point.